✨ implement amp-embedly-card component

juanlizarazo opened the pull request:
PR that implements `amp-embedly-card` component (feature request #12908) using [embedly cards.](http://docs.embed.ly/docs/cards)

juanlizarazo commented:
@aghassemi @alanorozco New PR ready for review :grin:

aghassemi commented:
Thanks @juanlizarazo.

I will defer the full review to @alanorozco 

On the API side: Is there a particular reason `amp-embedly-key` is a separate component? I suggest having a `data-key` attribute on the `<amp-embedly-card>` itself for consistency.

juanlizarazo commented:
@aghassemi 

It was a change requested on the original `<amp-embedly>`'s PR [here](https://github.com/ampproject/amphtml/pull/13843#discussion-diff-173265769R74). 

But maybe it made more sense then as it was required. For the new `<amp-embedly-card>` the paid api key is optional and it just removes branding.

I'll be happy to remove the additional `amp-embdely-key` component! Your call @aghassemi and @alanorozco

aghassemi commented:
@juanlizarazo I see. I don't have a strong opinion, leaving the final decision to @alanorozco on this.

alanorozco commented on extensions/amp-embedly-card/0.1/amp-embedly-card-impl.js:
```diff
@@ -0,0 +1,106 @@
+
+  /** @override */
+  layoutCallback() {
+    return getEmbedlyServiceForDoc(this.element).then(service => {
```
I know I originally suggested a service, but there's be a better way.

1. Remove the service
2. Simply query for 
```js
document.querySelector('amp-embedly-key').getAttribute('data-key');
```

The service suggestion was way overdesigned, I apologize.

I also read the other thread regarding the API. Unless you have an objection and after this change, it looks good to me!

juanlizarazo replied:
Thanks Alan, I will be updating this PR tonight :D

aghassemi commented:
Keeping this PR open since there is progress.

juanlizarazo commented on validator/testdata/feature_tests/amp-embedly-card.out:
```diff
@@ -0,0 +1 @@
+PASS
```
Updated!

juanlizarazo commented:
@alanorozco 

Updated as requested:

* Removed docservice to set/get embedly api key.
* Query value attribute on element as recommended. 
* Added validator test cases and updated .out file.

alanorozco commented on extensions/amp-embedly-card/amp-embedly-card.md:
````diff
@@ -0,0 +1,120 @@
+```
+
+Setting a key through the `amp-embedly-key` component is optional. If you are a paid user 
+this will remove Embedly's branding from the cards.
````
`This tag must only be included once`

juanlizarazo replied:
Updated

alanorozco commented on extensions/amp-embedly-card/validator-amp-embedly-card.protoascii:
```diff
@@ -0,0 +1,55 @@
+  }
+}
+
+tags: {  # <amp-embedly-key>
```
Add `unique: true`

juanlizarazo replied:
Added

honeybadgerdontcare commented on extensions/amp-embedly-card/validator-amp-embedly-card.protoascii:
```diff
@@ -0,0 +1,55 @@
+  requires_extension: "amp-embedly-card"
+  attrs: {
+    name: "data-url"
+    mandatory: true
```
You may want to consider adding some restrictions on the url. e.g. `https` only. See the [UrlSpec](https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L54) for what can be restricted. It can be set as a `value_url` on the tags spec, e.g.

```
value_url: {
  allowed_protocol: "https"
  allow_relative: false
}
````

juanlizarazo replied:
Added!

honeybadgerdontcare commented on extensions/amp-embedly-card/validator-amp-embedly-card.protoascii:
```diff
@@ -0,0 +1,55 @@
+tags: {  # <amp-embedly-card>
+  html_format: AMP
+  tag_name: "AMP-EMBEDLY-CARD"
+  satisfies: "amp-embedly-card"
```
This isn't needed (`satisfies`)

juanlizarazo replied:
Removed

honeybadgerdontcare commented on validator/testdata/feature_tests/amp-embedly-card.html:
```diff
@@ -0,0 +1,69 @@
+<!doctype html>
```
This should probably be in `extensions/amp-embedly-card/0.1/tests/` as `validator-amp-embedly-card.html` and the `.out` file as well.

juanlizarazo replied:
Moved to recommended folder and renamed.

alanorozco commented:
@bpaduch For assistance with docs.

juanlizarazo commented:
Awesome, thanks!

@alanorozco @bpaduch @aghassemi 

For the examples page in `examples/amp-embedly-card.amp.html`  should we remove the used key in `<amp-embedly-key>` just comment it out and have branded cards there?

As keys are paid, that key eventually will be eventually invalid.

alanorozco commented:
@juanlizarazo 

Sure, removing the key sounds good.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,118 @@
+  </tr>
+  <tr>
+    <td width="40%"><strong>Required Script</strong></td>
+    <td><code>&lt;script async custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-embedly-card-0.1.js">&lt;/script></code></td>
```
Change "amp-form" to "amp-embedly-card"

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,118 @@
+<table>
+  <tr>
+    <td width="40%"><strong>Description</strong></td>
+    <td>Provides you with responsive and shareable embeds to drive the reach of your websites, blog posts, and articles. from any URL using <a href="http://docs.embed.ly/docs/cards">Embedly cards</a>.</td>
```
remove period after articles

juanlizarazo replied:
Removed.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,116 @@
+*Example: Embedding multiple resources*
+
+If you have a paid plan, use the `amp-embedly-key` component to set your api key. 
+You just need one per document that includes one or multiple `amp-embedly-card` components:
```
need "one" what? If it's `amp-embedly-key`, please specify that.  

Like:

You just need one `amp-embedly-key` per AMP page. 


 Within the AMP page, you can specify one or multiple `amp-embedly-card` components:

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+*Example: Embedding multiple resources*
+
+If you have a paid plan, use the `amp-embedly-key` component to set your api key. 
+You just need one per AMP page.
```
Clarify what you mean by "one". For example "You just need one `amp-embedly-key` per AMP page.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
````diff
@@ -0,0 +1,114 @@
+</amp-embedly-key>
+```
+
+Setting a key through the `amp-embedly-key` component is optional. If you are a paid user, this removes Embedly's branding from the cards.
````
Would this be clearer: 

If you are a paid user, setting the `amp-embedly-key` removes Embedly's branding from the cards.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+
+##### data-card-via
+
+Specifies the via content in the card. This is a a great way to do attribution.
```
This isn't clear.  What's "via content" and what should be specified, a URL, an ID, etc.  Please specify: type, what it does, required or optional, and default value.

juanlizarazo replied:
Updated. 

Though I am not sure what this exactly does, it is not well documented. 

It has no impact on the card or its metadata. I contacted support to inquiry and clarify.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+
+##### data-card-chrome
+
+Removes the left-hand colored border if the value is set to `0`.
```
Is this a boolean or numeric value.  What's the possible values - please specify.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+
+##### data-card-theme
+
+For dark backgrounds it's better to specify the dark theme.
```
What are the possible values for this attribute.  Should also state what this attribute does, which is "Sets the card theme." then tell them what the default is, and possible values.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+
+##### data-card-image
+
+Specifies which image to use in article cards.
```
What should the value be - a URL, something else?  Please advise.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+
+##### data-card-controls
+	
+Enables share icons. The default value is `1`.
```
Clarify that by default share icons are enabled.  To turn them off, specify '0'.?

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+
+##### data-card-align
+
+Aligns the card. The default value is `"center"`.
```
What are the other values - please specify.

juanlizarazo replied:
Added possible values.

bpaduch commented on extensions/amp-embedly-card/amp-embedly-card.md:
```diff
@@ -0,0 +1,114 @@
+
+##### data-card-recommend
+
+Disables Embedly recommendations on video and rich cards. The default value is `1`.
```
Indicate that by default, recommendations are disabled (assuming "1" means that).

juanlizarazo replied:
Updated

juanlizarazo commented:
@bpaduch Updated with requested changes!

aghassemi commented:
Yay! Thanks @juanlizarazo 🎉

gctommy commented:
From the release notes ([https://github.com/ampproject/amphtml/releases/tag/1532461037969](https://github.com/ampproject/amphtml/releases/tag/1532461037969)) 5 days ago:

`✨ implement amp-embedly-card component`

But the script URL from the example ([https://cdn.ampproject.org/v0/amp-embedly-card-0.1.js](https://cdn.ampproject.org/v0/amp-embedly-card-0.1.js)) is not live yet?

aghassemi commented:
@juanlizarazo This is in Dev Channel right now (If you enabled Dev Channel in https://cdn.ampproject.org/experiments.html you can give it a try). Will make it to production over the next couple of days.

juanlizarazo commented:
Just tried it with different embed. Can't wait!

gctommy commented:
@aghassemi — [the documentation](https://www.ampproject.org/docs/reference/components/amp-embedly-card#attributes) should really mentioned that **data-url (required)** [only accepts https protocol](https://github.com/ampproject/amphtml/blob/master/extensions/amp-embedly-card/validator-amp-embedly-card.protoascii#L36), otherwise your page won't validate.

juanlizarazo commented:
@gctommy @aghassemi I can create a PR fixing that, I made sure the examples were compliant but I spaced out with the docs. 

I will fix a typo as well I saw there too and possibly remove the data-via property as well. I talked to embedly support about it and the sdk doesn't do what it is supposed to. So right now using `data-via` does nothing.
